### ISSUE TITLE
Example of using custom environments

### ISSUE THREAD

**Issue by UnnamedMoose** [HumanCompatibleAI/imitation#718]
## Problem
I have a custom environment derived from the Gym base class that works fine with stable baselines 3, including SubprocVecEnv, but I am struggling to integrate it with this framework because of the different wrapper layers.

## Solution
Having an example of a simple custom env (even a copy of on of the gym environments) would be very helpful to someone new to imitation like myself.

## Possible alternative solutions
A simple paragraph or two explaining how to use different environments instead of relying on the util::make_vec_env function would also be appreciated.

**Comment by ernestum** [comment]
Good point! To (maybe) unblock you for now I recommend to 
1. reigster you env using `gym.register()` with some unique ID
2. use `util.make_vec_env()` to create your environment with the given unique ID

**Comment by UnnamedMoose** [comment]
Thanks, I found a workaround by writing my own routine for creating the initial rollouts as that seemed to be the major issue when interfacing with my env as well as a PID controller that emulates a stable baselines model (apparently not well enough).  I'm pasting my code below as it may prove useful to others.

```
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas
import datetime
import os

import stable_baselines3
from stable_baselines3.common.vec_env import VecMonitor, SubprocVecEnv
from stable_baselines3.common.noise import NormalActionNoise, VectorizedActionNoise
import torch

from imitation.algorithms.adversarial.gail import GAIL
from imitation.algorithms.adversarial.airl import AIRL
from imitation.data.types import TrajectoryWithRew
from imitation.rewards.reward_nets import BasicRewardNet, BasicShapedRewardNet
from imitation.util.networks import RunningNorm

import verySimpleAuv as auv
import resources

font = {"family": "serif",
        "weight": "normal",
        "size": 16}

matplotlib.rc("font", **font)
matplotlib.rcParams["figure.figsize"] = (9, 6)

# %% Set up.
if __name__ == "__main__":
    # An ugly fix for OpenMP conflicts in my installation.
    os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

    agentName = "SAC_sblPretrain_try0_fromPID"

    # --- Pretraining parameters ---
    nPretrainEpisodes = 100
    nPretrainSteps = 500

    # Note that the bounds check is disabled to ensure episodes are of equal length.
    env_kwargs_pretrain = {
        "currentVelScale": 1.,
        "currentTurbScale": 2.,
        "noiseMagActuation": 0.,
        "noiseMagCoeffs": 0.,
        "stopOnBoundsExceeded": False,
    }

    # --- Training parameters ---

    # No. parallel processes.
    nProc = 1

    # Do everything N times to rule out random successes and failures.
    nAgents = 1

    # Any found agent will be left alone unless this is set to true.
    overwrite = True

    nTrainingSteps = 100_000
    agent_kwargs = {
        'learning_rate': 5e-4,
        'gamma': 0.95,
        'verbose': 1,
        'buffer_size': (64*3)*512,
        'batch_size': 256,
        'learning_starts': 256,
        'train_freq': (1, "step"),
        "gradient_steps": 1,
        # TODO how to use this with a different no processors?
        "action_noise": VectorizedActionNoise(NormalActionNoise(
            np.zeros(3), 0.05*np.ones(3)), nProc),
        "use_sde_at_warmup": False,
        "target_entropy": -4.,
    }
    policy_kwargs = {
        "activation_fn": torch.nn.GELU,
        "net_arch": dict(
            pi=[64, 64, 64],
            qf=[64, 64, 64],
        )
    }
    env_kwargs = {
        # Set to zero to disable the flow - much faster training.
        "currentVelScale": 1.,
        "currentTurbScale": 2.,
        # Use noise in coefficients for training only.
        "noiseMagActuation": 0.1,
        "noiseMagCoeffs": 0.1,
    }

    # --- Evaluation parameters ---
    makeAnimation = False

    env_kwargs_evaluation = {
        "currentVelScale": 1.,
        "currentTurbScale": 2.,
        "noiseMagActuation": 0.,
        "noiseMagCoeffs": 0.,
    }

# %% Create test rollouts.

    # Create an evaluation environment and Proportional-Derivative controller.
    env_eval = auv.AuvEnv(**env_kwargs_evaluation)
    pdController = auv.PDController(env_eval.dt)

    # Create an equivalent to imitation::util::crete_vec_env that instantiates
    # the custom environment.
    rng = np.random.default_rng(0)

    # Simply create the env without any wrappers.
    env_pretrain = auv.AuvEnv(**env_kwargs_pretrain)

    # Create own rollout generation to circumvent the issues with the imitation wrappers.
    # NOTE: this is where data could also be read from files, for instance.
    rollouts = []
    for iEp in range(60):
        obs = env_pretrain.reset()
        observations = [obs]
        actions = np.empty((0, env_pretrain.action_space.shape[0]))
        rewards = []
        for i in range(30):
            action, _states = pdController.predict(obs, deterministic=True)
            obs, reward, done, info = env_pretrain.step(action)
            observations = np.append(observations, obs[np.newaxis, :], axis=0)
            actions = np.append(actions, action[np.newaxis, :], axis=0)
            rewards = np.append(rewards, reward)

        out_dict_stacked = {"rews": rewards, "acts": actions, "obs": observations, "infos": None}
        traj = TrajectoryWithRew(**out_dict_stacked, terminal=done)
        assert traj.rews.shape[0] == traj.acts.shape[0] == traj.obs.shape[0] - 1

        rollouts.append(traj)

    # Pack the generated rollouts into DataFrames for eventual saving and easier visualisation.
    pretrainedEpisodes = []
    for r in rollouts:
        pretrainedEpisodes.append(pandas.DataFrame(
            data=np.concatenate([r.obs[1:, :], r.acts, r.rews[:, np.newaxis]], axis=1),
            columns=["s{:d}".format(i) for i in range(r.obs.shape[1])] +
                ["a{:d}".format(i) for i in range(r.acts.shape[1])] + ["r"]))

    # Plot the different training episodes.
    fig, ax = plt.subplots()
    ax.set_xlabel("s0")
    ax.set_ylabel("s1")
    for e in pretrainedEpisodes:
        cs = ax.scatter(e["s0"], e["s1"], c=e["r"], s=5)
    cbar = plt.colorbar(cs)
    cbar.set_label("Reward")

# %% Train the agents.

    # Create a parallel version of the environment.
    env_pretrain = SubprocVecEnv([auv.make_env(i, env_kwargs=env_kwargs_pretrain) for i in range(nProc)])

    # Train several times to make sure the agent doesn't just get lucky.
    convergenceData = []
    trainingTimes = []
    agents = []
    for iAgent in range(nAgents):
        # Set up constants etc.
        saveFile = "./agentData/{}_{:d}".format(agentName, iAgent)

        if not overwrite:
            if os.path.isfile(saveFile) or os.path.isfile(saveFile+".zip"):
                print("Skipping training of existing agent", saveFile)
                continue

        # Create the agent using stable baselines.
        agent = stable_baselines3.SAC(
            "MlpPolicy", env_pretrain, policy_kwargs=policy_kwargs, **agent_kwargs)

        # Evaluate
        # print("\nRandomly initialised agent")
        # resources.evaluate_agent(agent, env_eval, num_episodes=100)

        # Pretrain.
        reward_net = BasicRewardNet(
        # reward_net = BasicShapedRewardNet(
            env_pretrain.observation_space,
            env_pretrain.action_space,
            normalize_input_layer=RunningNorm,
        )
        pretrainer = GAIL(
        # pretrainer = AIRL(
            demonstrations=rollouts,
            demo_batch_size=1024,
            gen_replay_buffer_capacity=2048,
            n_disc_updates_per_round=4,
            venv=env_pretrain,
            gen_algo=agent,
            reward_net=reward_net,
            allow_variable_horizon=False,
            # log_dir="./tempData",
            # init_tensorboard=True,
        )
        pretrainer.train(nPretrainSteps)

        # Evaluate
        print("\nPretrained agent")
        resources.evaluate_agent(agent, env_eval, num_episodes=100)

        # Create and set a parallel enviornment for training
        env = SubprocVecEnv([auv.make_env(i, env_kwargs=env_kwargs) for i in range(nProc)])
        env = VecMonitor(env, saveFile)
        agent.set_env(env)

        # Train the agent for N more steps
        conv, trainingTime = resources.trainAgent(agent, nTrainingSteps, saveFile)
        convergenceData.append(conv)
        trainingTimes.append(trainingTime)
        agents.append(agent)

        # Evaluate
        print("\nTrained agent")
        resources.evaluate_agent(agent, env_eval, num_episodes=100)

        # Plot convergence of each agent. Redo after each agent to provide
        # intermediate updates on how the training is going.
        iBest, fig, ax = resources.plotTraining(
            convergenceData, saveAs="./agentData/{}_convergence.png".format(agentName))

    # Save metadata in human-readable format.
    resources.saveHyperparameteres(
        agentName, agent_kwargs, policy_kwargs, env_kwargs, nTrainingSteps, trainingTimes, nProc)
```

**Comment by AdamGleave** [comment]
Fixed in #746
